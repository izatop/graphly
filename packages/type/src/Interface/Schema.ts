import {InputObjectType, ObjectType} from "../Type";
import {IObject} from "./IObject";

export type TypeID = string | number;
export type TypeInt = number;
export type TypeFloat = number;
export type TypeString = string;
export type TypeBoolean = string;
export type TypeObject = { [key: string]: any };
export type TypeDate = Date;

export type OutputType<TObjectType extends ObjectType = ObjectType> = TObjectType
    | IObject
    | TypeID
    | TypeInt
    | TypeFloat
    | TypeString
    | TypeBoolean
    | TypeObject
    | TypeDate
    | string
    | number
    | boolean
    | undefined
    ;

export type InputType<TInputObjectType extends InputObjectType = InputObjectType> = TInputObjectType
    | TypeID
    | TypeInt
    | TypeFloat
    | TypeString
    | TypeBoolean
    | TypeObject
    | TypeDate
    | string
    | number
    | boolean
    | undefined
    ;
