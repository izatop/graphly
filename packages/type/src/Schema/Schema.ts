import {ObjectType, SubscriptionType} from "../Type";

export type SchemaCtor<T extends Schema = Schema> = {
    new(): T;
    getSchemaLocation(): NodeModule | string;
};

export abstract class Schema {
    public abstract readonly query: ObjectType;

    public readonly mutation?: ObjectType;

    public readonly subscription?: SubscriptionType;
}
