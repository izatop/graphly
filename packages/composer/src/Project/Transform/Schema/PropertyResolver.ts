import {XMap} from "@sirian/common";
import {
    GraphQLEnumType,
    GraphQLEnumValueConfigMap,
    GraphQLInputType,
    GraphQLList,
    GraphQLNonNull,
    GraphQLOutputType,
} from "graphql";
import {
    IPropertyFunction,
    IPropertyReference,
    ITypeEnum,
    ITypeObject,
    PropertyKind,
    PropertyType,
    ReturnType,
    TypeKind,
} from "../../../Type";
import {TYPE} from "../../../Type/const";
import {ucfirst} from "../../../util/ucfirst";
import {ScalarTypeTransform} from "../ScalarTypeTransform";
import {SchemaTransform} from "./SchemaTransform";

const cache = new XMap<string, any>();

export abstract class PropertyResolver<T extends GraphQLOutputType | GraphQLInputType> {
    public scalars = new ScalarTypeTransform();

    public context: SchemaTransform;

    public cache = cache;

    public base: Set<string>;

    constructor(context: SchemaTransform, base: Set<string>) {
        this.base = base;
        this.context = context;
    }

    public resolve(of: ITypeObject, property: PropertyType): T | undefined {
        if (property.kind === PropertyKind.SCALAR) {
            return this.scalars.transform(property.type) as T;
        }

        if (property.kind === PropertyKind.FUNCTION) {
            return this.resolveFunctionType(of, property);
        }

        if (property.kind === PropertyKind.REFERENCE) {
            return this.resolveReferenceType(of, property);
        }
    }

    protected resolveFunctionType(of: ITypeObject, property: IPropertyFunction): T | undefined {
        return this.resolve(of, property.returns);
    }

    protected resolveReferenceType(of: ITypeObject, property: IPropertyReference): T | undefined {
        if (this.scalars.has(property.reference)) {
            return this.scalars.transform(property.reference) as T;
        }

        if (this.cache.has(property.reference)) {
            return this.cache.ensure(property.reference);
        }

        if (ReturnType.has(property.reference)) {
            return this.resolve(of, property.parameters[0]);
        }

        if (property.reference === TYPE.ARRAY) {
            const type = this.resolve(of, property.parameters[0]);
            return type ? new GraphQLList(GraphQLNonNull(type)) as T : undefined;
        }

        if (this.context.types.has(property.reference)) {
            const type = this.context.types.ensure(property.reference);
            if (type.kind === TypeKind.CLASS) {
                this.cache.set(type.name, this.createObjectType(type));
            }

            if (type.kind === TypeKind.INTERFACE) {
                this.cache.set(type.name, this.createInterfaceType(
                    ucfirst(of.name) + ucfirst(property.name),
                    property,
                    type,
                ));
            }

            if (type.kind === TypeKind.ENUM) {
                this.cache.set(type.name, this.createEnumType(type));
            }
        }

        return this.cache.get(property.reference);
    }

    protected createEnumType(type: ITypeEnum): T {
        const values: GraphQLEnumValueConfigMap = {};
        for (const [key, value] of Object.entries(type.property)) {
            values[key] = {value};
        }

        return new GraphQLEnumType({
            name: type.name,
            values,
        }) as T;
    }

    protected abstract createObjectType(type: ITypeObject): T;

    protected abstract createInterfaceType(name: string, property: IPropertyReference, type: ITypeObject): T;
}
