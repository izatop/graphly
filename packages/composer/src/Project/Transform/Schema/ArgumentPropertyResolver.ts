import {GraphQLInputType} from "graphql";
import {IPropertyReference, ITypeObject, OutputType} from "../../../Type";
import {PropertyResolver} from "./PropertyResolver";
import {SchemaInputObjectTypeTransform} from "./SchemaInputObjectTypeTransform";
import {SchemaTransform} from "./SchemaTransform";

export class ArgumentPropertyResolver extends PropertyResolver<GraphQLInputType> {
    constructor(context: SchemaTransform) {
        super(context, OutputType);
    }

    protected createObjectType(type: ITypeObject) {
        return new SchemaInputObjectTypeTransform(this.context, type)
            .transform();
    }

    protected createInterfaceType(name: string, property: IPropertyReference, type: ITypeObject): GraphQLInputType {
        throw new Error(`Input types don't support interfaces`);
    }
}
