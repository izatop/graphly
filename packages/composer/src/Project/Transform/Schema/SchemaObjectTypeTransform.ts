import {ok} from "assert";
import {GraphQLFieldConfigMap, GraphQLObjectType, Thunk} from "graphql";
import {ITypeObject, OutputType} from "../../../Type";
import {TraceEvent} from "../../../util/TraceEvent";
import {TransformAbstract} from "../TransformAbstract";
import {SchemaObjectFieldTransform} from "./SchemaObjectFieldTransform";
import {SchemaTransform} from "./SchemaTransform";

type Args = [SchemaTransform, ITypeObject];

export class SchemaObjectTypeTransform extends TransformAbstract<Args, GraphQLObjectType> {
    protected traceEvent = TraceEvent.create(this);

    public get context() {
        return this.args[0];
    }

    public get type() {
        return this.args[1];
    }

    public get project() {
        return this.context.project;
    }

    public transform() {
        ok(
            OutputType.has(this.type.base),
            `Wrong object base type ${this.type.base} of ${this.type.name}`,
        );

        return new GraphQLObjectType({
            name: this.type.name,
            fields: () => this.fields(),
            // @todo Pick a description from comments
            description: `Description of ${this.type.name}`,
        });
    }

    protected fields(): GraphQLFieldConfigMap<any, any, any> {
        const fields: Thunk<GraphQLFieldConfigMap<any, any, any>> = {};
        for (const property of this.type.property) {
            fields[property.name] = new SchemaObjectFieldTransform(this.context, this, property)
                .transform();
        }

        return fields;
    }
}
