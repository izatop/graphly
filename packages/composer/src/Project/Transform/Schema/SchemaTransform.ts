import {ok} from "assert";
import {GraphQLInputType, GraphQLObjectType, GraphQLSchema, GraphQLSchemaConfig} from "graphql";
import {IProperty, IType, TypeBaseClass} from "../../../Serialization/interfaces";
import {Project} from "../../Project";
import {TransformAbstract} from "../TransformAbstract";
import {TypeTransform} from "../TypeTransform";
import {SchemaObjectTypeTransform} from "./SchemaObjectTypeTransform";

type Args = [Project];

export class SchemaTransform extends TransformAbstract<Args, GraphQLSchema> {
    public typeTransform = new TypeTransform();

    protected readonly objectType = new Map<string, GraphQLObjectType>();

    protected readonly inputType = new Map<string, GraphQLInputType>();

    public get project() {
        return this.args[0];
    }

    public get types() {
        return this.project.types;
    }

    public transform() {
        return new GraphQLSchema(this.transformSchemaType());
    }

    public hasInputType(name: string) {
        return this.inputType.has(name);
    }

    public getInputType(name: string) {
        return this.inputType.get(name)!;
    }

    public setInputType(name: string, type: GraphQLInputType) {
        this.inputType.set(name, type);
    }

    public hasObjectType(name: string) {
        return this.objectType.has(name);
    }

    public getObjectType(name: string) {
        return this.objectType.get(name)!;
    }

    public setObjectType(name: string, type: GraphQLObjectType) {
        this.objectType.set(name, type);
    }

    protected transformSchemaType(): GraphQLSchemaConfig {
        const schema = [...this.types.values()]
            .find(({base}) => base === TypeBaseClass.Schema)!;
        ok(schema, "Schema not found");

        const config: Partial<GraphQLSchemaConfig> = {};
        for (const property of schema.property) {
            Reflect.set(config, property.name, this.transformSchemaRoot(property));
        }

        ok(config.query, "The query field should be defined in a schema");
        return config as GraphQLSchemaConfig;
    }

    protected transformSchemaRoot(property: IProperty): GraphQLObjectType {
        ok(typeof property.type === "string");
        ok(this.types.has(property.type as string));

        const declaration = this.types.get(property.type as string)!;
        return new SchemaObjectTypeTransform(this, declaration)
            .transform();
    }
}
