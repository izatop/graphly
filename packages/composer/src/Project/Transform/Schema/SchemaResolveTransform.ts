import {ok} from "assert";
import {GraphQLFieldResolver} from "graphql";
import {resolve} from "path";
import {InputType, IPropertyFunction, PropertyKind, PropertyType, TypeKind} from "../../../Type";
import {TYPE} from "../../../Type/const";
import {TraceEvent} from "../../../util/TraceEvent";
import {TransformAbstract} from "../TransformAbstract";
import {SchemaObjectFieldTransform} from "./SchemaObjectFieldTransform";
import {SchemaTransform} from "./SchemaTransform";

type Args = [SchemaTransform, SchemaObjectFieldTransform, IPropertyFunction];
type Returns = GraphQLFieldResolver<any, any, any> | undefined;
type ResolveArgFn = (args: object, context: object) => any;

export class SchemaResolveTransform extends TransformAbstract<Args, Returns> {
    protected traceEvent = TraceEvent.create(this);

    public get schema() {
        return this.args[0];
    }

    public get context() {
        return this.args[1];
    }

    public get type() {
        return this.context.type;
    }

    public get property() {
        return this.args[2];
    }

    public transform(): Returns {
        const file = resolve(this.context.project.root, this.type.file);
        const module = require(file);
        ok(this.type.name in module, `Cannot find module ${this.type.name} at ${file}`);

        const owner: ObjectConstructor = Reflect.get(module, this.type.name);
        ok(
            this.property.name in owner.prototype,
            `Cannot find resolver ${this.type.name}.${this.property.name} at ${file}`,
        );

        const ns = `${this.type.name}.${this.property.name}`;
        const descriptor = Reflect.getOwnPropertyDescriptor(owner.prototype, this.property.name)!;
        const resolver = this.createMagickResolveFunction(ns, owner.prototype, descriptor);
        return function $resolve(parent: object, args: object, context: object) {
            return resolver(parent, args, context);
        };
    }

    protected createMagickResolveFunction(ns: string, proto: object, descriptor: PropertyDescriptor) {
        ok(
            descriptor && descriptor.value || descriptor.get,
            `Wrong resolver function ${this.type.name}.${this.property.name}`,
        );

        const map: ResolveArgFn[] = [];
        const resolver = descriptor.value || descriptor.get;
        for (const arg of this.property.args) {
            map.push(this.createArgResolver(arg));
        }

        return function $map$(parent: object, args: object, context: object) {
            const thisArg = Object.create(proto, Object.getOwnPropertyDescriptors(parent));
            const resolveArgs = map.map((fn) => fn(args, context));
            return resolver.apply(thisArg, resolveArgs);
        };
    }

    protected createArgResolver(property: PropertyType): ResolveArgFn {
        const {name} = property;
        const ns = `${this.type.name}.${this.property.name}`;
        if (property.kind === PropertyKind.SCALAR) {
            ok(
                InputType.has(property.type),
                `Wrong scalar type ${property.type} of ${ns}`,
            );

            return (args: {[k: string]: any}) => {
                return args[name];
            };
        }

        if (property.kind === PropertyKind.REFERENCE && property.parameters.length > 0) {
            ok(
                property.parameters.length === 1,
                `Wrong parameters count for ${property.reference} of ${ns}`,
            );

            return this.createArgResolver(property.parameters[0]);
        }

        if (property.kind === PropertyKind.REFERENCE) {
            if (this.schema.input.scalars.has(property.reference)) {
                return (args: {[k: string]: any}) => {
                    return args[name];
                };
            }

            if (this.schema.types.has(property.reference)) {
                const type = this.schema.types.ensure(property.reference)!;
                if (type.kind === TypeKind.CLASS) {
                    if (InputType.has(type.base)) {
                        return (args: {[k: string]: any}) => {
                            return args[name];
                        };
                    }
                }

                if (type.kind === TypeKind.SERVICE) {
                    if (type.base === TYPE.CONTEXT) {
                        return (args: {[k: string]: any}, context: object) => {
                            return context;
                        };
                    }

                    if (type.base === TYPE.CONTAINER) {
                        return (args: {[k: string]: any}, context: object) => {
                            return Reflect.get(context, "container");
                        };
                    }
                }
            }

            return () => {
                throw new Error(`Cannot resolve an argument type of ${ns}(${property.name}: ${property.reference})`);
            };
        }

        return () => {
            throw new Error(`Cannot resolve an argument type of ${ns}`);
        };
    }
}
