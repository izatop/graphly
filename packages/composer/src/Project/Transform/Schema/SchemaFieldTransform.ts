import {GraphQLFieldConfig, GraphQLFieldResolver, GraphQLList, GraphQLNonNull, GraphQLOutputType} from "graphql";
import * as vm from "vm";
import {IProperty, IPropertyResolver, IType, PropertyBox} from "../../../Serialization/interfaces";
import {TraceEvent} from "../../../util/TraceEvent";
import {Boxes} from "../interfaces";
import {TransformAbstract} from "../TransformAbstract";
import {SchemaObjectTypeTransform} from "./SchemaObjectTypeTransform";
import {SchemaResolveTransform} from "./SchemaResolveTransform";
import {SchemaTransform} from "./SchemaTransform";

type Args = [SchemaTransform, IType, IProperty | IPropertyResolver];
type Returns = GraphQLFieldConfig<any, any, any>;

export class SchemaFieldTransform extends TransformAbstract<Args, Returns> {
    protected traceEvent = TraceEvent.create(this);

    public get context() {
        return this.args[0];
    }

    public get type() {
        return this.args[1];
    }

    public get property() {
        return this.args[2];
    }

    public get types() {
        return this.context.types;
    }

    public get project() {
        return this.context.project;
    }

    public transform(): Returns {
        this.traceEvent.emit("transform", this.property);
        if (typeof this.property.type === "string") {
            return this.transformStringType();
        }

        if (this.property.type instanceof PropertyBox) {
            const type = this.transformBoxType(this.property.type);
            return {
                resolve: this.createResolver(),
                type: !this.property.nullable
                    ? new GraphQLNonNull(type)
                    : type,
            };
        }

        throw new Error(`Cannot resolve type of ${this.property.name}`);
    }

    protected transformBoxType(property: PropertyBox): GraphQLOutputType {
        const {type, args} = property;
        const [parameter, ...parameters] = args;

        if (parameter instanceof PropertyBox) {
            return this.transformBoxType(parameter);
        }

        if (type === Boxes.Array) {
            return new GraphQLList(this.resolveType(parameter as string));
        }

        return this.resolveType(parameter as string);
    }

    protected transformStringType(): Returns {
        const type = this.resolveType(this.property.type as string);

        return {
            resolve: this.createResolver(),
            type: !this.property.nullable
                ? new GraphQLNonNull(type)
                : type,
        };
    }

    protected resolveType(type: string) {
        this.traceEvent.emit("resolve", {
            name: this.type.name,
            scalar: this.context.typeTransform.has(type),
            objectType: this.context.types.has(type),
        });

        if (this.context.typeTransform.has(type)) {
            return this.context.typeTransform.transform(type);
        }

        if (this.context.types.has(type)) {
            return new SchemaObjectTypeTransform(this.context, this.context.types.get(type)!)
                .transform();
        }

        throw new Error(`Cannot resolve type of ${this.property.name}`);
    }

    protected createResolver(): GraphQLFieldResolver<any, any, any> | undefined {
        const hasResolver = "resolver" in this.property;
        const {defaultValue} = this.property;

        // A little hack to get a default value for a property which wasn't defined at start.
        // For example we have been added NewProp to a schema but can't update it on database
        // and want to sure it be an array whatever:
        // class Foo { public readonly myNewProp: NewPropType[] = [] }
        if (!hasResolver && defaultValue) {
            const propertyKey = this.property.name;
            const resolverFunction = function resolver(this: object) {
                return vm.runInContext(defaultValue, vm.createContext(this));
            };

            return (parent?: {[key: string]: any}) => {
                if (!parent || typeof parent[propertyKey] === "undefined") {
                    return resolverFunction.call(parent || {});
                }

                return parent[propertyKey];
            };
        }

        if (!hasResolver) {
            return;
        }

        return new SchemaResolveTransform(this.context, this.type, this.property).transform();
    }
}
