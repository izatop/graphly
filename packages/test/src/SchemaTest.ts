import {TodoStatus} from "@graphly/todo";
import {Scope} from "@graphly/type";
import * as assert from "assert";
import {ExecutionResult, graphql, parse, subscribe} from "graphql";
import {MainContainer} from "./MainContainer";
import {ITodo} from "./Repository/ITodo";
import {TestRepository} from "./Repository/TestRepository";
import {TodoInput} from "./Schema/Input/TodoInput";
import {TodoFlag} from "./Schema/Query/TodoFlag";
import {TestContext} from "./Schema/TestContext";
import {TestSchema} from "./TestSchema";

const todoFragment = `
fragment TodoFragment on Todo {
    id code title description status deadlineAt
    check { text solved } checklist { text solved }
}
`;

const todoTodoQuery = `
query TodoQuery ($id: Int!) {
    todo {
        todo(id: $id) { ...TodoFragment }
    }
}
${todoFragment}
`;

const todoTodoSearchQuery = `
query TodoSearchQuery {
    todo {
        search { node {...TodoFragment} }
    }
}
${todoFragment}
`;

const todoTodoSearchAllQuery = `
query TodoSearchQueryAll {
    todo {
        searchAll { node {...TodoFragment } }
    }
}
${todoFragment}
`;

const todoCountQuery = `
query TodoCountQuery {
    todo {
        count
    }
}
`;

const addTodoQuery = `
mutation addTodo($todo: TodoInput!) {
    todo { add(todo: $todo) { ...TodoFragment } }
}
${todoFragment}
`;

const updateTodoQuery = `
mutation updateTodo($id: Int! $todo: TodoInput!) {
    todo { update(id: $id todo: $todo) { ...TodoFragment } }
}
${todoFragment}
`;

const onTodoUpdate = `
subscription subscribeTodoUpdate {
    onTodoUpdate { id flag title checklist }
}
${todoFragment}
`;

const isAsyncIterable = (source: ExecutionResult | AsyncIterable<ExecutionResult>):
    source is AsyncIterableIterator<ExecutionResult> => {
    return Symbol.asyncIterator in source;
};

describe("Composer", () => {
    const config = {
        dsn: "connection string",
    };

    const scope = new Scope({
        schema: TestSchema,
        context: TestContext,
        container: MainContainer,
        config,
    });

    const runQuery = async (q: string, v?: Record<string, any>): Promise<ExecutionResult> => {
        const state = {timestamp: Date.now(), authorized: false, session: ""};
        const factory = await scope.create(() => state);
        const {schema, contextValue, rootValue} = await factory(undefined);

        return graphql({
            schema,
            rootValue,
            contextValue,
            variableValues: v,
            source: q,
        });
    };

    const runSubscribe = async (q: string, v?: Record<string, any>): Promise<ExecutionResult | AsyncGenerator> => {
        const state = {timestamp: Date.now(), authorized: false, session: ""};
        const factory = await scope.create(() => state);
        const {schema, contextValue, rootValue} = await factory(undefined);

        return subscribe({
            schema,
            rootValue,
            contextValue,
            document: parse(q),
            variableValues: v,
        });
    };

    test("Test Subscribe", async () => {
        const todo: TodoInput = {
            checklist: [],
            deadlineAt: new Date(),
            title: "New Todo",
            status: TodoStatus.PENDING,
        };

        const repository = new TestRepository(config.dsn);
        const todos = repository.get<ITodo>("todos");
        const iterator = await runSubscribe(onTodoUpdate) as AsyncGenerator<any>;

        const {id} = await todos.add(todo);
        await todos.update(id, {flag: TodoFlag.PRIVATE});

        const items = [];
        assert.ok(isAsyncIterable(iterator));
        for await (const item of iterator) {
            expect(item.errors).toBeUndefined();
            if (item.errors) {break;}
            if (item.data?.onTodoUpdate.id === id) {
                items.push(item);
                break;
            }
        }

        expect(items.length).toBe(1);
        expect(items).toMatchSnapshot();
    });

    test("Test Context", async () => {
        const state = {timestamp: Date.now(), authorized: false, session: ""};
        const factory = await scope.create(() => state, {test: true});
        const {contextValue, rootValue} = await factory(undefined);
        expect(contextValue.container.repository).toBeInstanceOf(TestRepository);
        expect(contextValue.container.config).toMatchObject(config);
        expect(contextValue.container.config).toBe(contextValue.getConfig());
        expect(contextValue.state).toMatchObject(state);
        expect(rootValue).toMatchObject({test: true});
    });

    test("Schema query", async () => {
        const query = "query {optional random timestamp hello}";
        const {data} = await runQuery(query);
        expect(data?.optional).toBe(null);
        expect(typeof data?.random).toBe("number");
        expect(typeof data?.timestamp).toBe("number");
        expect(data?.hello).toBe("Hello world");

        const todo: TodoInput = {
            checklist: [],
            deadlineAt: new Date(),
            title: "New Todo",
            status: TodoStatus.PENDING,
        };

        const add = await runQuery(addTodoQuery, {todo});
        expect(add.errors).toBeUndefined();
        expect(add.data).toMatchObject({
            todo: {
                add: {
                    ...todo,
                    id: 1,
                    deadlineAt: todo.deadlineAt?.getTime(),
                    description: null,
                },
            },
        });

        const todoQuery = await runQuery(todoTodoQuery, {id: 1});
        const {repository} = await scope.createContainer();
        const firstTodo = repository.get("todos").findOne(1);

        expect(todoQuery.errors).toBeUndefined();
        expect(todoQuery.data).toMatchObject({
            todo: {
                todo: firstTodo,
            },
        });

        const updateTodo: TodoInput = {
            ...todo,
            checklist: [{text: "Checklist 1", solved: true}],
            description: "Description",
            deadlineAt: new Date(),
            status: TodoStatus.DONE,
        };

        const update = await runQuery(updateTodoQuery, {id: 1, todo: updateTodo});
        expect(update.errors).toBeUndefined();
        expect(update.data).toMatchObject({
            todo: {
                update: {
                    ...updateTodo,
                    deadlineAt: updateTodo.deadlineAt?.getTime(),
                    id: 1,
                },
            },
        });

        const nullableQueryTest = await runQuery(todoTodoQuery, {id: 999});
        expect(nullableQueryTest.errors).toBeUndefined();
        expect(nullableQueryTest.data).toMatchObject({todo: {todo: null}});

        const todoCount = await runQuery(todoCountQuery);
        expect(todoCount.errors).toBeUndefined();
        expect(todoCount.data).toMatchObject({todo: {count: 1}});
    });

    const iPageableTestQuery = [todoTodoSearchQuery, todoTodoSearchAllQuery];
    test.each(iPageableTestQuery)("IPageable", async (q) => {
        const query = await runQuery(q);
        expect(query.errors).toBeUndefined();
    });
});
