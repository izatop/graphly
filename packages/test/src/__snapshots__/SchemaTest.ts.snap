// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Composer Test Subscribe 1`] = `
Array [
  Object {
    "data": Object {
      "onTodoUpdate": Object {
        "checklist": Array [],
        "flag": "PUBLIC",
        "id": 1,
        "title": "New Todo",
      },
    },
  },
]
`;
